parser/openfoodfacts_taxonomy_parser/parser/parser.py

```diff
@@ -1,158 +1,213 @@
-                if key.startswith("prop_"):
-                    entry_query += " SET n." + key + " = $" + key + "\n"
+            raise ValueError(f"ENTRY node type should be batched")
 
+        entry_query = ""
         for key in node_data.tags:
-            if key.startswith("tags_"):
-                entry_query += " SET n." + key + " = $" + key + "\n"
+            entry_query += " SET n." + key + " = $" + key + "\n"
 
         query = id_query + entry_query + position_query
-        self.session.run(query, node_data.to_dict())
+        tx.run(query, node_data.to_dict())
 
-    def _get_project_name(self, taxonomy_name: str, branch_name: str):
-        """Create a project name for given branch and taxonomy"""
-        return "p_" + taxonomy_name + "_" + branch_name
+    def _create_entry_nodes(self, entry_nodes: list[NodeData], project_label: str):
+        """Create all ENTRY nodes in a single batch query"""
+        self.parser_logger.info("Creating ENTRY nodes")
+        start_time = timeit.default_timer()
 
-    def _create_multi_label(self, taxonomy_name: str, branch_name: str) -> str:
-        """Create a combined label with taxonomy name and branch name"""
-        project_name = self._get_project_name(taxonomy_name, branch_name)
-        return project_name + ":" + ("t_" + taxonomy_name) + ":" + ("b_" + branch_name)
-
-    def create_nodes(self, nodes: list[NodeData], multi_label: str):
-        """Adding nodes to database"""
-        self.parser_logger.info("Creating nodes")
-        for node in nodes:
-            if node.id == "__header__":
-                self._create_headernode(node.preceding_lines, multi_label)
-            else:
-                self._create_node(node, multi_label)
+        base_query = f"""
+          WITH $entry_nodes as entry_nodes
+          UNWIND entry_nodes as entry_node
+          CREATE (n:{project_label}:ENTRY)
+          SET n.id = entry_node.id
+          SET n.preceding_lines = entry_node.preceding_lines
+          SET n.src_position = entry_node.src_position
+          SET n.main_language = entry_node.main_language
+        """
+        additional_query = ""
+        seen_properties = set()
+        seen_tags = set()
+        for entry_node in entry_nodes:
+            if entry_node.get_node_type() != NodeType.ENTRY:
+                raise ValueError(f"Only ENTRY nodes can be batched")
+            for key in entry_node.properties:
+                if not key in seen_properties:
+                    additional_query += " SET n." + key + " = entry_node." + key + "\n"
+                    seen_properties.add(key)
+            for key in entry_node.tags:
+                if not key in seen_tags:
+                    additional_query += " SET n." + key + " = entry_node." + key + "\n"
+                    seen_tags.add(key)
+
+        query = base_query + additional_query
+        self.session.run(query, entry_nodes=[entry_node.to_dict() for entry_node in entry_nodes])
```

Comment:
I'm quite positively surprised by the flexibility of Neo4j to run this even if there are properties is missing in some nodes :-)